source/extensions/clusters/aggregate/cluster.cc

```diff
@@ -0,0 +1,153 @@
+  Upstream::PrioritySetImpl priority_set;
+  std::vector<std::pair<uint32_t, Upstream::ThreadLocalCluster*>> priority_to_cluster;
+  int next_priority_after_linearizing = 0;
+
+  // Linearize the priority set. e.g. for clusters [C_0, C_1, C_2] referred in aggregate cluster
+  //    C_0 [P_0, P_1, P_2]
+  //    C_1 [P_0, P_1]
+  //    C_2 [P_0, P_1, P_2, P_3]
+  // The linearization result is:
+  //    [C_0.P_0, C_0.P_1, C_0.P_2, C_1.P_0, C_1.P_1, C_2.P_0, C_2.P_1, C_2.P_2, C_2.P_3]
+  // and the traffic will be distributed among these priorities.
+  for (const auto& cluster : clusters_) {
+    if (skip_predicate(cluster)) {
+      continue;
+    }
+    auto tlc = cluster_manager_.get(cluster);
+    if (tlc == nullptr) {
+      continue;
+    }
+
+    int priority_in_current_cluster = 0;
+    for (const auto& host_set : tlc->prioritySet().hostSetsPerPriority()) {
+      if (!host_set->hosts().empty()) {
+        priority_set.updateHosts(
+            next_priority_after_linearizing++, Upstream::HostSetImpl::updateHostsParams(*host_set),
+            host_set->localityWeights(), host_set->hosts(), {}, host_set->overprovisioningFactor());
+        priority_to_cluster.emplace_back(std::make_pair(priority_in_current_cluster, tlc));
+      }
+      priority_in_current_cluster++;
+    }
+  }
+
+  return std::make_pair(std::move(priority_set), std::move(priority_to_cluster));
+}
+
+void Cluster::startPreInit() {
+  for (const auto& cluster : clusters_) {
+    auto tlc = cluster_manager_.get(cluster);
+    if (tlc == nullptr) {
+      continue;
+    }
+
+    // Add callback for clusters initialized before aggregate cluster.
+    tlc->prioritySet().addMemberUpdateCb(
+        [this, cluster](const Upstream::HostVector&, const Upstream::HostVector&) {
+          ENVOY_LOG(info, "member update for cluster '{}'", cluster);
+          refresh();
+        });
+  }
+  refresh();
+  handle_ = cluster_manager_.addThreadLocalClusterUpdateCallbacks(*this);
+
+  onPreInitComplete();
+}
+
+void Cluster::refresh(std::function<bool(const std::string&)>&& skip_predicate) {
+  auto pair = linearizePrioritySet(std::move(skip_predicate));
+  // Post the priority set to worker threads.
+  tls_->runOnAllThreads([this, &pair, cluster_name = this->info()->name()]() {
+    LoadBalancerImplSharedPtr lb_ = nullptr;
+    if (!pair.first.hostSetsPerPriority().empty()) {
+      lb_ = std::make_shared<LoadBalancerImpl>(*this, std::move(pair));
+    }
+    Upstream::ThreadLocalCluster* cluster = cluster_manager_.get(cluster_name);
+    if (cluster == nullptr) {
+      return;
+    }
+    // Downgrade cast to AggregateClusterLoadBalancer
+    dynamic_cast<AggregateClusterLoadBalancer&>(cluster->loadBalancer()).refresh(lb_);
+  });
+}
+
+void Cluster::onClusterAddOrUpdate(Upstream::ThreadLocalCluster& cluster) {
+  if (std::find(clusters_.begin(), clusters_.end(), cluster.info()->name()) != clusters_.end()) {
+    ENVOY_LOG(info, "update or add cluster '{}'", cluster.info()->name());
```

Comment:
same comment here: make it clearer that we're talking about the aggregate cluster